packages/test-utils/src/run.ts

```diff
@@ -72,24 +77,31 @@ function gatherEntities(result: any): Entity[] {
   }
 }
 
-async function mapResultToOriginalEm<R>(em: EntityManager, result: R): Promise<R> {
+async function mapResultToOriginalEm<R>(em: EntityManager, result: R, opts?: RunOptions): Promise<R> {
   const newEmEntities = gatherEntities(result);
   // load any entities that don't exist in the original em
   await Promise.all(
     newEmEntities
-      .filter(
-        (e) =>
-          !em.findExistingInstance(
-            e.idTaggedMaybe ?? fail("Joist 'run' returned an unsaved entity; are you missing an em.flush?"),
-          ),
-      )
+      .filter((e) => {
+        if (e.idTaggedMaybe === undefined) {
+          if (opts?.expectUnsavedEntities === true) return false;
+          fail("Joist 'run' returned an unsaved entity; are you missing an em.flush?");
+        }
+        return !em.findExistingInstance(e.idTaggedMaybe);
+      })
       .map((e) => em.load(e.idTagged)),
   );
   // generate a cache of id -> entity in original em
-  const cache = Object.fromEntries(newEmEntities.map((e) => [e.id, em.findExistingInstance(e.idTagged) as Entity]));
+  const cache = Object.fromEntries(
+    newEmEntities
+      // Exclude entities which have not been flushed, so that they don't fail the lookup
+      .filter((e) => isDefined(e.idTaggedMaybe))
+      .map((e) => [e.id, em.findExistingInstance(e.idTagged) as Entity]),
+  );
   function doMap(value: any): any {
     if (isEntity(value)) {
-      return cache[value.id];
+      // If the value has no ID, simply return it as is
+      return value.idMaybe ? cache[value.id] : value;
```

Comment:
issue: this seems problematic as it's effectively breaking the `em` isolation that is the entire point of `run`.  This should really be a clone from the old `em` to the new.   Also as pointed out previously, it would be possible for an entity to have an id at this point but not be in the cache (if the previous issue was fixed).